카드 순차적 출력 구현, ViewController 코드 정리

TTOzzi opened the pull request:
말씀해주신 대로 한 장씩 출력이 아닌 여러 장씩 출력되게 하니 카드 장수가 적어서 그런지 오차가 적게 출력되었습니다. 여전히 출력 중일 때 게임 모드를 바꾸거나 셔플을 하면 조금 꼬여서 출력되긴 하지만 이 방법의 한계인 것 같습니다.
ViewController 에서 View 관련 코드를 최대한 분리해주었습니다.
서로에게 영향을 주는 오토레이아웃과 모델의 데이터를 넘기는 것을 제외하고 분리하니 코드가 생각만큼 깔끔하진 않은 것 같습니다.

TTOzzi committed:
스트레이트 확인 로직 수정

값을 반환하여 가져가서 비교하는 방식에서 값을 객체 내부로 가져와서 다음 카드인지 확인하는 방법으로 수정

TTOzzi committed:
getNumbers() 삭제

굳이 함수를 만들어놓을 이유가 없는것같아 삭제

TTOzzi committed:
변경사항에 맞게 코드 수정

TTOzzi committed:
불필요 테스트코드 제거

TTOzzi committed:
점수 테스트 코드 작성

TTOzzi committed:
불필요 코드 삭제

TTOzzi committed:
CardImageView 분리

TTOzzi committed:
HandStackView 분리

TTOzzi committed:
PlayerStackView 분리

TTOzzi committed:
플레이어 라벨 생성 로직 변경

Player 에 number 프로퍼티를 추가하고 CustomString 을 활용해서 그에 맞게 라벨이 생성되도록 해줌

TTOzzi committed:
View 관련 파일들 View 폴더로 이동

TTOzzi committed:
SegmentedControl 분리

TTOzzi committed:
GameView, WinnerImageView 분리

TTOzzi committed:
SegmentedControl 구조 변경

값이 바뀔 때 호출하는 함수를 하나로 통일

TTOzzi committed:
게임 재시작시 카드가 부족하면 덱을 리셋하도록 수정

TTOzzi committed:
코드 양식 수정

TTOzzi committed:
카드 순차적 출력 구현

godrm commented on CardGameApp/CardGameApp/Model/Card.swift:
```diff
@@ -28,9 +30,11 @@ class Card {
+        if self.rank.isNextRank(with: card.rank) {
+            return true
+        }
+        return false
```
이 정도는 비교문을 사용하지 말고 그냥 return self.rank.isNextRank(with: card.rank) 형태로 써도 좋습니다.

TTOzzi replied:
앗 그걸 생각 못했네요 바로 수정하겠습니다

godrm commented on CardGameApp/CardGameApp/Model/Hand.swift:
```diff
@@ -22,7 +22,7 @@ class Hand {
     }
     private lazy var score = getScore()
-    private lazy var numbers = getNumbers()
+    private lazy var ranks = getInfo().map { (value) -> Card in value.key }
```
getInfo().map() 표현이 명확하게 어떤 의도인지 모르겠네요. 
좀 더 구체적으로 동작을 표현하는 메소드 이름을 표현하면 좋겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/Hand.swift:
```diff
@@ -34,23 +34,19 @@ class Hand {
     private func checkStraight() -> Bool {
-        let sortedCards = cards.sorted { $0 > $1 }
+        let sortedCards = cards.sorted { $0 < $1 }
+        var currentCard = sortedCards[0]
```
checkStraight() 보다는 isStraight() 처럼 check-는 명확하지 않은 표현이니까 더 명확한 표현을 사용하세요.

TTOzzi committed:
불필요 코드 삭제

godrm commented on CardGameApp/CardGameApp/Model/Hand.swift:
```diff
@@ -97,25 +93,20 @@ class Hand {
                     continue
                 } else {
-                    return lhs.numbers[index] > rhs.numbers[index]
+                    return lhs.ranks[index] > rhs.ranks[index]
```
이 부분이 아쉽네요. ranks 탐색을 여기서 하지말고, ranks 내부에 구현해서 ranks 끼리 비교하는 어떤가요?

TTOzzi replied:
여기서 ranks 가 Hand 에서 score 가 같을 때 승자를 구분하기위해 score 가 높은순으로 숫자들을 정렬한 배열인데 Hand 의 extension 에서 Hand 의 프로퍼티를 비교하는것이라 상관없을거라 생각했습니다. 그러면 이것도 따로 클래스를 만들어줘야할까요..?

godrm replied:
데이터를 추상화하는 것은 새로운 데이터 타입을 만드는 게 시작입니다. 
다른 타입을 배열에 넣고 사용하면, 그 배열을 추가, 삭제, 탐색, 정렬하고 비교하는 코드가 계속 늘어나게되죠. 
그런 것을 요약해서 다른 곳에서도 쓸 수 있도록 추상화하는 시도를 계속해보세요. 
las.ranks > rhs.ranks 이렇게 비교할 수 있으면 더 간결하고 좋은거죠.

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -9,8 +9,13 @@
     private var hand = Hand()
     
+    init(number: Int? = nil) {
+        self.number = number
```
왜 Int? 가 됐나요? 
Dealer를 이렇게 특정한 값으로 구분하는 것은 좋지 않습니다.
동일한 구현을 하고 더 구체적으로 딜러의 동작만 한다면 Dealer를 상속받아 구현하는 방법도 있습니다.
숫자에 따라 다르게 동작하는 코드가 있으면 좋은 구조가 아닙니다.

TTOzzi replied:
Dealer 클래스가 Player 클래스를 상속받는 구조로 구현해주었고 이 number 부분은 수정하기 전에는 뷰컨트롤러에서 라벨을 만들 때 플레이어의 번호를 지정해주기위해 index 변수를 만들어 따로 만들어줬었는데 그걸 여기서 처리해주기위해 추가해주었습니다. 이것으로 딜러와 플레이어를 구분해주진 않습니다.

TTOzzi replied:
이 부분을 Dealer 에 따로 CustomStringConvertible 을 채택해주어 해결하려했는데 Dealer 가 Player 를 상속받아서 중복 적용이라 적용이 되지 않습니다. 그냥 원래 했던 방식대로 VC 에서 index 를 처리해줘야할까요??

godrm replied:
중복적용? 프로토콜 채택한걸 다시 구현하는게 안 되던가요?

TTOzzi replied:
Dealer 가 Player 를 상속받은 상태에서 Player 에 CustomStringConvertible 이 채택되어있습니다. 여기서 Dealer 에 따로 CustomStringConvertible 을 채택해주려니 이미 채택한 프로토콜이라고 적용이 안되네요. override 도 안되구요.

godrm replied:
이런 구조가 되지 않나요? 
```
class Player : CustomStringConvertible {
    var description: String {
        "player"
    }
}

class Dealer : Player {
    override var description: String {
        "dealer"
    }
}
```

TTOzzi replied:
우와 따로 extension 으로 빼서 프로토콜을 채택해줬었는데 클래스 안에 넣으니 되네요!
extension 으로 했을땐 왜 안됐을까요..?

godrm replied:
이미 extension 한걸 다시 바꾸는건 안됩니다. 상속 관계를 처리하려면 클래스 안에서 해야해요.

TTOzzi replied:
아 그렇군요
감사합니다!!

godrm commented on CardGameApp/CardGameApp/Model/Player.swift:
```diff
@@ -24,7 +29,14 @@ class Player {
+        guard let number = number else {
+            return "Dealer"
+        }
+        return "Player\(number)"
```
여기 있네요. 이렇게 값에 따라 다른 동작을 해야 한다면 프로토콜 또는 상속으로 클래스를 분리하는 게 좋습니다.

TTOzzi replied:
이 부분은 같이 처리하지않고 따로 딜러에 프로토콜을 채택해주는게 더 좋을 것 같네요. 수정하겠습니다.

godrm commented on CardGameApp/CardGameApp/Model/PokerGame.swift:
```diff
@@ -48,11 +48,11 @@ class PokerGame {
         self.dealer = Dealer()
         numberOfPlayers.forEach {
-            self.players.add(player: Player())
+            self.players.add(player: Player(number: players.count + 1))
```
이 부분도 게임이라면 players.add 를 직접하는 게 아니라 players.addPlayers(number of:numberOfPlayers)  형태로 간단하게 호출하고, Players 가 여기 numberOfPlayers.forEach {}를 갖고 있는 게 좋습니다. 
이렇게 상위 모듈은 하위 모듈에 한 줄로 일을 시킬 수 있어야 합니다. 참고하세요.

TTOzzi committed:
getRanks 함수 추가

godrm commented on CardGameApp/CardGameApp/View/GameView.swift:
```diff
@@ -0,0 +1,34 @@
+    
+    required init(coder: NSCoder) {
+        super.init(coder: coder)
+        self.translatesAutoresizingMaskIntoConstraints = false
```
지금은 한 줄일 뿐이지만, 두 init 에서 공통적인 코드를 넣기 보다는 하나의 지정 생성자를 호출하도록 하는 게 좋습니다.

godrm commented on CardGameApp/CardGameApp/View/PlayerStackView.swift:
```diff
@@ -0,0 +1,41 @@
+    
+    convenience init(playerName: String) {
+        self.init()
+        addPlayerLabel(playerName: playerName)
```
스위프트에서 init() 을 호출하는 것은 아무것도 초기화되지 않아서 동작하지 않을 경우가 많습니다. 
뷰의 지정 생성자를 꼭 호출하는 게 좋습니다. 참고하세요.

TTOzzi committed:
checkStraight 함수명 변경

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -54,125 +41,64 @@ class ViewController: UIViewController {
+                    self.addWinnerMedal(to: self.pokerGame.winner)
+                }
+            }
+        }
```
애니메이션에 대한 부분은 추후에 다시 한 번 얘기할 기회가 있을 겁니다.

godrm commented on CardGameApp/ScoreTest/ScoreTest.swift:
```diff
@@ -0,0 +1,63 @@
+        hand.add(card: Card(shape: .spade, rank: .eight))
+        XCTAssertEqual(hand.score, .onePair)
+    }
+}
```
다른 테스트 경우는 없을까요? straight 끼리 비교나 Pair끼리 비교도 있으면 좋겠습니다.

TTOzzi replied:
넵 추가해보겠습니다!

TTOzzi committed:
Ranks 클래스 생성

TTOzzi committed:
Score 클래스 생성

TTOzzi committed:
플레이어 추가를 PokerGame 이 아닌 Players 에서 처리해주도록 수정

TTOzzi committed:
CustomStringConvertible 프로토콜 수정

extension 은 상속시 프로토콜 오버라이드가 안됨
상속받아 바꿔야할 프로토콜의 경우 클래스 내부에 구현하기

TTOzzi committed:
플레이어 추가 수정

플레이어를 추가 할 때 딜러를 생성해서 넣으니 포커게임의 딜러와 추가한 딜러가 다름
포커게임의 딜러를 받아 추가하도록 수정

TTOzzi committed:
코드 양식 수정

TTOzzi committed:
Straight 일때 비교 로직 수정

페어일때는 Ranks 로 비교가 가능한데 스트레이트일땐 현재 로직으론 Ranks 로 비교가 불가능함
Straight 일땐 straightNumber 를 만들어 비교해줌

TTOzzi committed:
두 플레이어간 비교 테스트 추가

TTOzzi committed:
View 생성 시 중복코드 setUp() 함수로 분리

TTOzzi commented:
말씀해주신 내용들 반영하여 수정 완료했습니다!
상속받은 프로토콜을 오버라이드하기 위해선 상위클래스에서 extension 이 아닌 클래스 안에 구현해야한다는것을 새롭게 알게되어 좋았습니다:+1:

godrm commented:
수고하셨습니다 🤖
